src/context/WalletProvider/useEip1993EventHandler.ts

```diff
@@ -0,0 +1,124 @@
+      // This fixes the case of switching from e.g MM, to another wallet, then switching accounts/chains in MM and MM becoming connected again
+      if (_walletType && localWalletType !== _walletType) return
+      if (!localWalletType) return
+
+      const _isLocked = Array.isArray(accountsOrChains) && accountsOrChains.length === 0
+
+      if (_isLocked) {
+        dispatch({ type: WalletActions.SET_IS_LOCKED, payload: true })
+      } else {
+        // Either a chain change or a wallet unlock - ensure we set isLocked to false before continuing to avoid bad states
+        dispatch({ type: WalletActions.SET_IS_LOCKED, payload: false })
+      }
+
+      const adapter = (await getAdapter(localWalletType)) as MetaMaskAdapter | PhantomAdapter | null
+
+      // Re-pair - which in case of accounts changed means the user will be prompted to connect their current account if they didn't do so
+      // Note, this isn't guaranteed to work, not all wallets are the same, some (i.e MM) have this weird flow where connecting to an unconnected account
+      // from a connected account can only be done from the wallet itself and not programmatically
+      const localWallet = await adapter?.pairDevice()
+
+      if (!localWallet) return
+
+      await localWallet.initialize()
+      const deviceId = await localWallet?.getDeviceID()
+
+      if (!deviceId) return
+
+      const { icon, name } = SUPPORTED_WALLETS[localWalletType]
+
+      dispatch({
+        type: WalletActions.SET_WALLET,
+        payload: {
+          wallet: localWallet,
+          name,
+          icon,
+          deviceId,
+          connectedType: localWalletType,
+        },
+      })
+    },
+    [dispatch, getAdapter, localWalletType, maybeMipdProvider, state.adapters],
+  )
+
+  const setProviderEvents = useCallback(() => {
+    // Always remove before setting
+    maybeMipdProvider?.provider.removeListener?.('accountsChanged', handleAccountsOrChainChanged)
+    maybeMipdProvider?.provider.removeListener?.('chainChanged', handleAccountsOrChainChanged)
+
+    maybeMipdProvider?.provider.on?.('accountsChanged', (e: string[]) => {
+      return handleAccountsOrChainChanged(e)
+    })
+    maybeMipdProvider?.provider.on?.('chainChanged', (e: string) => {
+      return handleAccountsOrChainChanged(e)
+    })
+  }, [handleAccountsOrChainChanged, maybeMipdProvider?.provider])
+
+  // Register a MetaMask-like (EIP-1193) provider's event handlers on mipd provider change
+  useEffect(() => {
+    const isMetaMaskMultichainWallet = state.wallet instanceof MetaMaskMultiChainHDWallet
+    const isPhantomHdWallet = state.wallet instanceof PhantomHDWallet
+    if (!(isMetaMaskMultichainWallet || isPhantomHdWallet)) return
+    if (!maybeMipdProvider) return
+    try {
+      setProviderEvents()
+    } catch (e) {
+      if (!isMobile) console.error(e)
+    }
+  }, [maybeMipdProvider, setProviderEvents, state.wallet])
+}
```

Comment:
q: don't you think there is a race-condition risk where the `provider` could be replaced before the `removeListeners` are called and so the hold listeners are still existing?

Reply:
There may be and this is expected, this is the actual intent here: `removeListener` is more of a paranoia than anything else. The fix here in making things work is the `rdns` `useRef()` - ensuring we always tap into the right provider, without all the previous `getMaybeProvider`, window.ethereum, and `removeAllProviderListeners` madness on each wallet's `pairDevice()` call